src/macros.jl

```diff
@@ -166,3 +166,92 @@ end
 @inline (self::SkipOptimize{F})(a) where {F} = self.f(value(a))
 @inline (self::SkipOptimize{F})(a, b) where {F} = self.f(value(a), value(b))
 @inline (self::SkipOptimize{F})(a, b, c) where {F} = self.f(value(a), value(b), value(c))
+
+"""
+    f(x) = dot(x, x)
+    f(x::ReverseDiff.TrackedVector) = ReverseDiff.track(f, x)
+    ReverseDiff.@grad function f(x)
+        xv = ReverseDiff.value(x)
+        return dot(xv, xv), ∇ -> (∇ * 2 * xv,)
+    end
+
+The `@grad` macro provides a way for the users to define custom adjoints for single-output functions wrt to their input numbers or arrays.
+"""
+macro grad(expr)
+    if @capture(expr, 
+        (f_(xs__) where {T__} = body_) | 
+        (f_(xs__) = body_) | 
+        (function f_(xs__) body_ end) | 
+        (function f_(xs__) where {T__} body_ end)
+    )
```

Comment:
For breaking appart function bodies I suggest using ExprTools.jl's `splitdef`
there are a bunch of edge cases around this, turns out there are a lot of AST possibilities for how a function can be declared, and it handles all of them, robustly.

Reply:
Ya in DistributionsAD, I ended up using MacroTools to support kwargs as well.